### ISSUE TITLE
Unified API for conversion

### ISSUE THREAD

**Issue by wangjiawen2013** [cpmech/plotpy#74]
Hi,
I am performing boxplot recently days and have two questions about the conversion functions. You have defined five functions:
```
/// Converts vector to a Python list of numbers
pub(crate) fn vector_to_numbers<T>(buf: &mut String, name: &str, vector: &[T])

/// Converts vector to a Python list of strings
pub(crate) fn vector_to_strings<T>(buf: &mut String, name: &str, vector: &[T])

/// Converts vector to a 1D NumPy array
pub(crate) fn vector_to_array<'a, T, U>(buf: &mut String, name: &str, vector: &'a T)

/// Converts a matrix to a nested Python list
pub(crate) fn matrix_to_list<T>(buf: &mut String, name: &str, matrix: &Vec<Vec<T>>)

/// Converts a matrix to a 2D NumPy array
pub(crate) fn matrix_to_array<'a, T, U>(buf: &mut String, name: &str, matrix: &'a T)
```
As the above shows, when converting the data to Python list, you used `&[T]` in the signatures, while it's `&'a T` when converting the data to np.array.  Why don't change `&[T]` to `&'a T` to make the functions more general and unified such as both by column? If so, we can also convert rust 1dvector and 1darray to python list and convert rust 2d slice and 2d array to python 2d array too.

And besides, Now `matrix_to_list`  converts the elements by row, but `matrix_to_array` converts by column, they are transposed. It's expected they should give the same results. Such as for `let x = vec![vec![1,2,3,4,5], vec![2,3,4,5,6], vec![3,4,5,6,7]];` . The reults of `matrix_to_list` is `[[1,2,3,4,5], [2,3,4,5,6], [3,4,5,6,7]]`, while it is `[[1,2,3], [2,3,4], [3,4,5], [4,5,6], [5,6,7]]` for `matrix_to_array`. Matplotlib both supports python list and numpy array/ pandas dataframe, but the former conversion will plot three boxes when performing boxplot using matplotlib, the later conversion will plot five boxes ! Converting both by row or by column for the two functions will exclude the discrepency and make plotpy API more unified.

![捕获](https://github.com/user-attachments/assets/6822a171-dd1b-4abb-a8b6-ff0224457468)

**Comment by cpmech** [comment]
Hi, interesting question! We actually need both versions of each function. I renamed three functions to clarify things:

* `generate_list`: generates a Python list, given a Rust slice
* `generate_list_quoted:` generates a Python list with entries singly-quoted, given a Rust slice
* `generate_nested_list`: generates a list of lists in Python, given a nested Rust "Vec" (not actually a vector, but an array, mathematically speaking)

The other functions remain with the same name:

* `vector_to_array`: converts an "actual vector" (mathematically speaking) to a 1D numpy array
* `matrix_to_array`: converts an "actual matrix" (mathematically speaking) to a 2D numpy array

The difference in the ordering of nested lists versus matrices is intentional. In the nested list, a "row" can be "longer" than the next row. Matrices are exactly "rectangular".

Internally, we can pick the conversion function that generates the data format we expect. This does not affect the user; see the Histogram for example: https://docs.rs/plotpy/latest/plotpy/struct.Histogram.html

**Comment by cpmech** [comment]
Hi again; I have updated my previous comment with code samples.

**Comment by cpmech** [comment]
I have also run the test with your numbers:

```rust
#[test]
fn nested_list_versus_matrix() {
    let a = vec![
        vec![1, 2, 3, 4, 5], //
        vec![2, 3, 4, 5, 6], //
        vec![3, 4, 5, 6, 7], //
    ];
    let mut nested_list = String::new();
    let mut numpy_array = String::new();
    generate_nested_list(&mut nested_list, "a_python", &a);
    matrix_to_array(&mut numpy_array, "a_numpy", &a);
    println!("{}", nested_list);
    println!("{}", numpy_array);
}
```

Results:

```text
a_python=[[1,2,3,4,5,],[2,3,4,5,6,],[3,4,5,6,7,],]

a_numpy=np.array([[1,2,3,4,5,],[2,3,4,5,6,],[3,4,5,6,7,],],dtype=float)
```

And:

```rust
#[test]
fn nested_list_versus_matrix() {
    let a = vec![
        vec![1, 2, 3, 4, 5],     //
        vec![5, 6, 7, 8, 9],     //
        vec![9, 10, 11, 12, 13], //
    ];
    let mut nested_list = String::new();
    let mut numpy_array = String::new();
    generate_nested_list(&mut nested_list, "a_python", &a);
    matrix_to_array(&mut numpy_array, "a_numpy", &a);
    println!("{}", nested_list);
    println!("{}", numpy_array);
}
```

Results:

```text
a_python=[[1,2,3,4,5,],[5,6,7,8,9,],[9,10,11,12,13,],]

a_numpy=np.array([[1,2,3,4,5,],[5,6,7,8,9,],[9,10,11,12,13,],],dtype=float)
```

The results are as expected.

**Comment by cpmech** [comment]
Another thought: it may be that your data had fewer entries in the first row. In this case, `matrix_to_array` ignores the extra entries in the subsequent rows. For example (from https://github.com/cpmech/plotpy/blob/main/src/as_matrix.rs):

```rust
#[test]
fn as_matrix_works() {
    // heap-allocated 2D array (vector of vectors)
    const IGNORED: f64 = 123.456;
    let a = vec![
        vec![1.0, 2.0],
        vec![3.0, 4.0, IGNORED, IGNORED, IGNORED],
        vec![5.0, 6.0],
    ];
    assert_eq!(
        matrix_str(&a),
        "1,2,\n\
         3,4,\n\
         5,6,\n"
    );
}
```

Since the matrix must be rectangular, if you need such "ignored values", you need to use the `generate_nested_list` function.

**Comment by wangjiawen2013** [comment]
As `generate_nested_list` only accepts nested vector as parameter:
`pub(crate) fn generate_nested_list<T>(buf: &mut String, name: &str, data: &Vec<Vec<T>>)`
If I want to convert an 2d array to python list, I must convert the 2d array to a nested vector before
`let d = [[1,2,3].to_vec(),[1,2,3].to_vec()].to_vec();`
Could `generate_nested_list` also support 2d array directly ? Is so, this function will be more general and convinent.

**Comment by cpmech** [comment]
Unfortunately, Rust does not support such a feature unless we develop some new trait. And things may get complicated from there. Because of such limitation, the Histogram only accepts a `Vec<Vec<...>>`:

```rust
pub fn draw<T, U>(&mut self, values: &Vec<Vec<T>>, labels: &[U]) {...}
```

Does your function accept different input formats? i.e., not `Vec<Vec<...>>`?

**Comment by wangjiawen2013** [comment]
I'll restrict the parameter to accept only nested vectors.

**Comment by cpmech** [comment]
Hello, thanks again for your help. I've implemented both options: with a nested list and with a 2D array---available in **1.8.0**. See the examples below (and in the documentation):

## Data as a nested list

```rust
use plotpy::{Boxplot, Plot, StrError};

fn main() -> Result<(), StrError> {
    // data (as a nested list)
    let data = vec![
        vec![1, 2, 3, 4, 5],              // A
        vec![2, 3, 4, 5, 6, 7, 8, 9, 10], // B
        vec![3, 4, 5, 6],                 // C
        vec![4, 5, 6, 7, 8, 9, 10],       // D
        vec![5, 6, 7],                    // E
    ];

    // x ticks and labels
    let ncol = data[0].len();
    let ticks: Vec<_> = (1..(ncol + 1)).into_iter().collect();
    let labels = ["A", "B", "C", "D", "E"];

    // boxplot object and options
    let mut boxes = Boxplot::new();
    boxes.draw(&data);

    // save figure
    let mut plot = Plot::new();
    plot.add(&boxes)
        .set_title("boxplot documentation test")
        .set_ticks_x_labels(&ticks, &labels)
        .save("/tmp/plotpy/doc_tests/doc_boxplot_2.svg")?;
    Ok(())
}
```

![doc_boxplot_2.svg](https://raw.githubusercontent.com/cpmech/plotpy/main/figures/doc_boxplot_2.svg)

## Data as a 2D array

```rust
use plotpy::{Boxplot, Plot, StrError};

fn main() -> Result<(), StrError> {
    // data (as a 2D array/matrix)
    let data = vec![
        //   A  B  C  D  E
        vec![1, 2, 3, 4, 5],
        vec![2, 3, 4, 5, 6],
        vec![3, 4, 5, 6, 7],
        vec![4, 5, 6, 7, 8],
        vec![5, 6, 7, 8, 9],
        vec![6, 7, 8, 9, 10],
        vec![14, 14, 14, 14, 14], // outliers
    ];

    // x ticks and labels
    let ncol = data[0].len();
    let ticks: Vec<_> = (1..(ncol + 1)).into_iter().collect();
    let labels = ["A", "B", "C", "D", "E"];

    // boxplot object and options
    let mut boxes = Boxplot::new();
    boxes.draw_mat(&data);

    // save figure
    let mut plot = Plot::new();
    plot.add(&boxes)
        .set_title("boxplot documentation test")
        .set_ticks_x_labels(&ticks, &labels)
        .save("/tmp/plotpy/doc_tests/doc_boxplot_1.svg")?;
    Ok(())
}
```

![doc_boxplot_1.svg](https://raw.githubusercontent.com/cpmech/plotpy/main/figures/doc_boxplot_1.svg)

**Comment by cpmech** [comment]
I did some further testing/illustration:

```rust
let data = vec![
    //    A    B    C    D    E ← matrix: columns are series
    //                            nested: rows are series
    //                                ↓
    vec![1.0, 2.0, 3.0, 4.0, 5.0], // A
    vec![1.1, 2.1, 3.1, 4.1, 5.1], // B
    vec![1.2, 2.2, 3.2, 4.2, 5.2], // C
    vec![1.3, 2.3, 3.3, 4.3, 5.3], // D
    vec![1.4, 2.4, 3.4, 4.4, 5.4], // E
];

let ticks = [1, 2, 3, 4, 5];
let labels = ["A", "B", "C", "D", "E"];

let mut boxes_nes = Boxplot::new();
boxes_nes.draw(&data);

let mut boxes_mat = Boxplot::new();
boxes_mat.draw_mat(&data);

let mut plot = Plot::new();
plot.set_subplot(1, 2, 1)
    .set_title("nested")
    .add(&boxes_nes)
    .set_ticks_x_labels(&ticks, &labels)
    .set_subplot(1, 2, 2)
    .set_title("matrix")
    .add(&boxes_mat)
    .set_ticks_x_labels(&ticks, &labels);
```

![boxplot](https://raw.githubusercontent.com/cpmech/plotpy/main/figures/integ_boxplot_3.svg)

Thus, both functions work as expected, according to Matplotlib's specification.
